eclipse-collections-code-generator/src/main/resources/impl/list/immutable/immutablePrimitiveArrayList.stg

```diff
@@ -699,6 +705,38 @@ final class Immutable<name>ArrayList
         throw new UnsupportedOperationException("subList not yet implemented!");
     }
 
+    /**
+     * @since 9.1.
+     */
+    @Override
+    public ImmutableList\<<name><name>Pair> zip<name>(<name>List list)
+    {
+        MutableList\<<name><name>Pair> target = Lists.mutable.empty();
+        int size = this.size();
+        int othersize = list.size();
+        for (int i = 0; i \< size && i \< othersize; i++)
+        {
+            target.add(PrimitiveTuples.pair(this.items[i], list.get(i)));
+        }
+        return target.toImmutable();
+    }
+
+    /**
+     * @since 9.1.
+     */
+    @Override
+    public \<T> ImmutableList\<<name>ObjectPair\<T\>> zip(ListIterable\<T> list)
+    {
+        MutableList\<<name>ObjectPair\<T\>> target = Lists.mutable.empty();
```

Comment:
We should have a shortcut which checks for list.isEmpty() and returns. I am fine merging this PR if you want to do it as a part of next commit/PR. Please raise an issue and add "good for first issue" label.

Reply:
I don't think it's necessary to special case isEmpty(). However, calling list.get() without checking RandomAccess makes this O(n^2) worst case.

Reply:
Good point.